### ISSUE TITLE
P1-21: TrainSkeletonSpec 作成前の training assembly 共通型を導入する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#513]
---
title: "P1-21: TrainSkeletonSpec 作成前の training assembly 共通型を導入する"
labels: ["refactor", "P1", "codex"]
---

# P1-21: TrainSkeletonSpec 作成前の training assembly 共通型を導入する

## 依存 issue

01

## 対象ファイル

- `packages/seisai-engine/src/seisai_engine/pipelines/common/training_assembly.py（新規）`
- `packages/seisai-engine/src/seisai_engine/pipelines/common/__init__.py`
- `packages/seisai-engine/tests/test_training_assembly.py（新規）`
- `packages/seisai-engine/tests/test_train_skeleton_loop.py`

## 背景

pair/psn/blindtrace train は最終的に `TrainSkeletonSpec` を作って `run_train_skeleton` を呼ぶが、その前段の model/dataset/criterion/optimizer assembly が各 main に散らばっている。

## 目的

`PipelineBuildResult` と `make_train_skeleton_spec` などの共通型・関数を追加し、pipeline ごとの factory 化の足場を作る。

## 非対象

- pair/psn/blindtrace の main 移行は後続 issue で行う。
- `TrainSkeletonSpec` 自体の field rename はしない。
- training loop の挙動変更はしない。

## 実装手順

1. `PipelineBuildResult` dataclass を作る。候補 field: `model`, `model_sig`, `optimizer`, `criterion_train`, `criterion_eval`, `ds_train_full`, `ds_infer_full`, `infer_epoch_fn`, `ckpt_extra`, `vis_adapter`。実際の既存 spec に必要な範囲に絞る。
2. `make_train_skeleton_spec(common, cfg, base_dir, build_result, ...)` を作るか、より小さく `build_spec_from_parts` とする。
3. ckpt validation 共通ルール（`save_best_only`, `metric`, `mode`）を helper 化する。
4. tile/infer worker 固定など既存 common validation と重複する場合は既存 helper を呼ぶ。
5. unit test では dummy model/dataset/callable を使い、spec field mapping と validation error を確認する。

## 受け入れ条件

- training assembly tests が通る。
- 既存 train skeleton tests が通る。
- common module lazy exports に必要最小限の symbol が追加される。
- 既存 train modules はまだ未移行でも壊れない。

## 推奨テスト

```bash
PYTHONPATH=packages/seisai-engine/src:packages/seisai-utils/src:packages/seisai-dataset/src:packages/seisai-transforms/src:packages/seisai-models/src:packages/seisai-pick/src \
python -m pytest \
  packages/seisai-engine/tests/test_training_assembly.py \
  packages/seisai-engine/tests/test_train_skeleton_loop.py \
  packages/seisai-engine/tests/test_train_skeleton_ckpt_payload.py
```

```bash
PYTHONPATH=packages/seisai-engine/src:packages/seisai-utils/src:packages/seisai-dataset/src:packages/seisai-transforms/src:packages/seisai-models/src:packages/seisai-pick/src \
python -m pytest packages/seisai-engine/tests/test_import_smoke.py packages/seisai-engine/tests/test_lazy_import_boundaries.py
```
